server/lib/vision_tagger.py

```
@@ -182,31 +239,58 @@ def tag_image(image_url: str, *, prompt: str | None = None) -> tuple[dict, float
+        )
+        usage = getattr(response, "usage_metadata", None)
+
+        parsed: VisionTagResult | None = getattr(response, "parsed", None)
+        if parsed is None:
+            try:
+                parsed = VisionTagResult.model_validate_json(response.text or "")
+            except ValidationError as e:
+                error_label = "schema_invalid"
+                raise RuntimeError(
+                    "Gemini response did not match VisionTagResult schema."
+                ) from e
```

Ruff (0.15.12)

[warning] Avoid specifying long messages outside the exception class

(TRY003)